src/caches/abstractbeliefpropagationcache.jl

```diff
@@ -144,6 +152,28 @@ function incoming_messages(
   return incoming_messages(bpc, [partition_vertex]; kwargs...)
 end
 
+#Adapt interface for changing device
+function adapt_messages(to, bpc::AbstractBeliefPropagationCache)
+  bpc = copy(bpc)
+  for pe in keys(messages(bpc))
+    set_message!(bpc, pe, adapt(to).(message(bpc, pe)))
+  end
+  return bpc
+end
+function adapt_factors(to, bpc::AbstractBeliefPropagationCache)
+  bpc = copy(bpc)
+  for v in vertices(bpc)
+    @preserve_graph bpc[v] = adapt(to).(bpc[v])
+  end
+  return bpc
+end
```

Comment:
Maybe these could be written more generally as `map_messages` and `map_factors` that take any function, and then `adapt_structure` could be written in terms of `map_messages(adapt(to), bpc)`, etc.